pkg/agent/multicast/multicast_route.go

```diff
@@ -0,0 +1,422 @@
+	origin := src.To4()
+	mc.Origin = [4]byte{origin[0], origin[1], origin[2], origin[3]}
+	g := group.To4()
+	mc.Mcastgrp = [4]byte{g[0], g[1], g[2], g[3]}
+	mc.Parent = iif
+	return multicastsyscall.SetsockoptViMfcctl(c.sockFD, syscall.IPPROTO_IP, multicastsyscall.MRT_DEL_MFC, mc)
+}
+
+func (c *MRouteClient) flushMRoute() {
+	klog.InfoS("Clearing multicast routing table entries")
+	multicastsyscall.SetsockoptVifctl(c.sockFD, syscall.IPPROTO_IP, multicastsyscall.MRT_FLUSH, &multicastsyscall.Vifctl{})
+}
+
+func (c *MRouteClient) deleteInboundMrouteEntryByGroup(group net.IP) (err error) {
+	klog.V(2).InfoS("Deleting multicast group", "group", group)
+	mEntries, _ := c.inboundRouteCache.ByIndex(GroupNameIndexName, group.String())
+	for _, route := range mEntries {
+		entry := route.(*inboundMulticastRouteEntry)
+		err := c.delMrouteEntry(net.ParseIP(entry.src), net.ParseIP(entry.group), entry.vif)
+		if err != nil {
+			return err
+		}
+		c.inboundRouteCache.Delete(route)
+	}
+	return nil
+}
+
+// addOutBoundMrouteEntry configures multicast route from Antrea gateway to oifs(outbound multicast interfaces)
+// to allow multicast senders in the pods send multicast traffic to external.
+func (c *MRouteClient) addOutBoundMrouteEntry(src net.IP, group net.IP, oifs []string) (err error) {
+	klog.InfoS("Adding outbound multicast route", "src", src, "group", group, "oifs", oifs)
+	vifs := make([]uint16, 0, len(oifs))
+	for _, oif := range oifs {
+		oifvif, err := c.vifAllocator.setVifIfNotExist(oif)
+		if err != nil {
+			return err
+		}
+		vifs = append(vifs, oifvif)
+	}
+
+	inboundVif, _ := c.vifAllocator.getVif(c.gatewayInterface)
+	err = c.addMrouteEntry(src, group, uint16(inboundVif), vifs)
+	if err != nil {
+		return err
+	}
+	routeEntry := &outboundMulticastRouteEntry{
+		group: group.String(),
+		src:   src.String(),
+		vifs:  vifs,
+	}
+	_, exist, _ := c.outboundRouteCache.GetByKey(routeEntry.group + "/" + routeEntry.src)
+	if !exist {
+		c.outboundRouteCache.Add(routeEntry)
+	} else {
+		c.outboundRouteCache.Update(routeEntry)
+	}
+	return nil
+}
+
+// addInboundMrouteEntry configures multicast route from external interface to Antrea gateway
+// to allow multicast receivers in the pods receive multicast traffic from external.
+func (c *MRouteClient) addInboundMrouteEntry(src net.IP, group net.IP, inboundVif uint16) (err error) {
+	klog.InfoS("Adding inbound multicast route", "src", src, "group", group, "inboundVif", inboundVif)
+	oif, err := c.vifAllocator.getVif(c.gatewayInterface)
+	if err != nil {
+		return err
+	}
+
+	err = c.addMrouteEntry(src, group, inboundVif, []uint16{oif})
+	if err != nil {
+		return err
+	}
+	routeEntry := &inboundMulticastRouteEntry{
+		group: group.String(),
+		src:   src.String(),
+		vif:   inboundVif,
+	}
+	c.inboundRouteCache.Add(routeEntry)
+	return nil
+}
+
+// inboundMulticastRouteEntry encodes the inbound multicast routing entry.
+// e.g. (10.0.0.55,226.94.9.9)           Iif: wlan0      Oifs: antrea-gw0
```

Comment:
I cannot associate the example with the fields in the struct. Is there any change?

Reply:
I have updated the comments. Hope it is more readable.